Move FQN related code into a separate NuGet package

Haplois opened the pull request:
## Description
As part of https://github.com/microsoft/vstest/issues/2708:
 - [x] Moved ManagedNameUtilities into a new project.
 - [x] Removed hardcoded support from TestCase class to restore compatibility with older versions of TestPlatform.
 - [x] Create a new NuGet package
 - [x] Determine what to call `Microsoft.TestPlatform.ManagedNames` package. We're calling it `Microsoft.TestPlatform.AdapterUtilities`.
 - [x] Fix Ubuntu build

Haplois committed:
Added `.svclog` into gitignore

Haplois committed:
Removed ManagedNameUtilities from ObjectModel.

Removed ManagedNameUtilities from ObjectModel in order to restore compatibility with older versions.

Haplois committed:
Added `Microsoft.TestPlatform.ManagedNames.nuspec`.

 - NuGet package recipe for `Microsoft.TestPlatform.ManagedNames` added.

Haplois committed:
Fixed a typo in resource access.

Haplois committed:
Publish test results on failed builds.

Haplois committed:
Fixed failed tests.

Haplois committed:
Merge branch 'master' into fqn-separation

Sanan07 commented on src/Microsoft.TestPlatform.ManagedNames/Extensions/ReflectionExtensions.cs:
```diff
@@ -1,20 +1,20 @@
             if (obj == null)
             {
-                throw new NotImplementedException(string.Format(Resources.MethodNotImplementedOnPlatform, className, methodName));
+                throw new NotImplementedException(string.Format(ManagedNameMessages.MethodNotImplementedOnPlatform, className, methodName));
```
Better to change to ManagedNameResources to be consistent with other resources.

Haplois replied:
You're right, but this resource is moved from `ObjectModel`, and it's already translated. It's unnecessarily long refactoring for no apparent gain IMO. (All `.resx`, `.xlf` and generated code needs to be updated)

Sanan07 commented on src/Microsoft.TestPlatform.ManagedNames/ManagedNameUtilities/Contants.cs:
```diff
@@ -0,0 +1,14 @@
+        public const string ManagedTypeLabel = "ManagedType";
+        public const string ManagedMethodLabel = "ManagedMethod";
+
+        public const string ManagedTypePropertyId = "TestCase." + ManagedTypeLabel;
```
Why do not have value equal to "TestCase.ManagedType" as it is const?

Sanan07 commented on src/Microsoft.TestPlatform.ManagedNames/Extensions/TestCase2Extensions.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Returns <c>true</c> if both <see cref="ITestCase2.ManagedType" /> and <see cref="ITestCase2.ManagedMethod" /> are not null or whitespace.
+        /// </summary>
+        public static bool HasManagedMethodAndType(this ITestCase2 testCase)
```
```suggestion
        public static bool ContainsManagedMethodAndType(this ITestCase2 testCase)
```

Haplois committed:
PR recommendations applied.

cvpoienaru commented on src/Microsoft.TestPlatform.ManagedNames/Extensions/TestCase2Extensions.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Returns <c>true</c> if both <see cref="ITestCase2.ManagedType" /> and <see cref="ITestCase2.ManagedMethod" /> are not null or whitespace.
+        /// </summary>
+        public static bool HasManagedMethodAndType(this ITestCase2 testCase)
```
Assuming we cannot change the `ITestCase2` interface, maybe we should consider introducing `ITestCase3` that inherits `ITestCase2` and adds the new methods/properties instead of working around with extension methods ?

Haplois replied:
`ITestCase2` interface is new. `ContainsManagedMethodAndType` is not included there; I'll include it. The interface is explicitly implemented, so these extension methods provide ease of access to those members.

cvpoienaru commented on src/Microsoft.TestPlatform.ManagedNames/ManagedNameUtilities/ManagedNameParser.cs:
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.ManagedNameUtilities
 
     public class ManagedNameParser
```
Maybe this class should be internal too ?

Haplois replied:
We like to provide these helper methods mainly for adapter authors. That's the reason they are public.

cvpoienaru commented on src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:
```diff
@@ -450,6 +329,35 @@ protected override void ProtectedSetPropertyValue(TestProperty property, object
+        /// <inheritdoc />
+        string ITestCase2.GetDisplayName()
+        {
+            if (this.HasManagedMethodAndType())
```
nit: This method's body can be re-written using the ternary operator, but I acknowledge it's a matter of personal taste.

Haplois replied:
Method removed.

peterwald commented on src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:
```diff
@@ -10,13 +10,13 @@ namespace Microsoft.VisualStudio.TestPlatform.ObjectModel
     /// </summary>
     [DataContract]
-    public sealed class TestCase : TestObject
+    public sealed class TestCase : TestObject, ITestCase2
```
It seems like there could be a failure in the case where a new test adapter (implementing the Managed properties), was used with an older version of the test platform.  Say, for instance, that vstest.console is run from the command line on an older install, but it is referencing a new adapter that supports the new properties. I think it would build and load properly, but the extension methods that access ITestCase2 would fail. Can we protect against that?

Haplois replied:
The problem is that we can't access `TestProperty` from this new package. The only way I can see is not providing this interface at all and doing everything in the test adapters; in that case, I don't see a way to customize `TestId` generation. Any ideas?

peterwald replied:
If we do everything in the test adapters, it seems to me that would work. TestId's are ideally created in the test adapter, so it seems like they could use the new code in the nuget package to generate it. Am I missing something?

Haplois replied:
> TestId's are ideally created in the test adapter

In this case, no. I'll move the functionality into the test adapter.

Haplois committed:
Merged with master

Haplois committed:
Fixed non-windows builds.

Haplois committed:
Removed ManagedNames from TestCase

Haplois committed:
Package renamed to `Microsoft.TestPlatform.AdapterUtilities`.

 - All references to `Microsoft.TestPlatform.ManagedNames` refactored as `Microsoft.TestPlatform.AdapterUtilities`.

shyamnamboodiripad commented:
@Haplois @peterwald Can we now delete the Reflection-based implementation of FQN from the VSUnitTesting repo?

I would still like to ensure that the Reflection and Roslyn based implementations continue to align - so it would be great to preserve unit tests for the FQN functionality in VSUnitTesting repo (i.e. replace the Reflection implementation testted by the unit tests with the new MS.TP.AdapterUtilities nuget - but continue to reference Roslyn based implementation from the VSUnitTesting repo).

shyamnamboodiripad commented on src/Microsoft.TestPlatform.AdapterUtilities/Microsoft.TestPlatform.AdapterUtilities.csproj:
```diff
@@ -0,0 +1,53 @@
+  
+  <PropertyGroup>
+    <TargetFrameworks>netstandard1.0;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">$(TargetFrameworks);uap10.0;</TargetFrameworks>
```
Just an observation on something that is likely to come up in the future: Since managed test frameworks will probably require this reference to be added in their framework layer (as opposed to the VS / adapter layer) and since the framework layer will likely be consumed in older versions of VS (as well as outside of VS), test frameworks may end up needing to conditionally add package reference for `MS.TP.AdapterUtilities` to only a subset of their supported targets and / or conditionally opt in to the FQN functionality in their framework layer only in cases where they know that the caller is VS. Alternatively, if they can't consume the package they may need to maintain a parallel implementation of the FQN spec or copy the Reflection code from the current repo...

shyamnamboodiripad reviewed (commented):
👍🏾 modulo comments

Haplois commented on src/Microsoft.TestPlatform.AdapterUtilities/ManagedNameUtilities/ManagedNameHelper.Reflection.cs:
```diff
@@ -139,7 +139,7 @@ public static MethodBase GetManagedName(Assembly assembly, string fullTypeName,
 
             if (type == null)
```
I refactored the parameters on the first method, as you suggested. However, for the second method (one that returns MethodBase), the argument names were actually misleading. `fullTypeName` is actually the type name of `Assembly.GetName`, and not our `ManagedTypeName`. I renamed `fullTypeName` to `typeName`, and `fullMethodName` to `managedMethodName` on that method.

Haplois commented on src/Microsoft.TestPlatform.AdapterUtilities/Microsoft.TestPlatform.AdapterUtilities.csproj:
```diff
@@ -0,0 +1,53 @@
+
+    <PackageReference Include="System.Runtime.InteropServices.RuntimeInformation" Version="4.3.0" />
+    <PackageReference Include="System.Diagnostics.FileVersionInfo" Version="4.3.0" />
+    <PackageReference Include="Microsoft.Win32.Primitives" Version="4.3.0" />
```
The last time I used this, it was called something else (win8 era :)). I don't know where these come from. I copied those from another project. I will try to remove and compile.

Haplois commented on src/Microsoft.TestPlatform.AdapterUtilities/Microsoft.TestPlatform.AdapterUtilities.csproj:
```diff
@@ -0,0 +1,53 @@
+  
+  <PropertyGroup>
+    <TargetFrameworks>netstandard1.0;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">$(TargetFrameworks);uap10.0;</TargetFrameworks>
```
TP only supports net451 and newer. It looks like netstandard1.0 maps into net45 (https://dotnet.microsoft.com/platform/dotnet-standard#versions). I added net20 and net35 support. The code is uglier now :)

AbhitejJohn commented on scripts/build.ps1:
```diff
@@ -805,16 +815,19 @@ function Create-NugetPackages
+    $nuspecFiles = @(
+        "TestPlatform.TranslationLayer.nuspec",
+        "TestPlatform.ObjectModel.nuspec",
+        "Microsoft.TestPlatform.AdapterUtilities.nuspec",
```
nit nit nit: could we order them on name?

Haplois replied:
done.

AbhitejJohn commented on src/Microsoft.TestPlatform.AdapterUtilities/Extensions/ReflectionExtensions.MethodBase.cs:
```diff
@@ -1,12 +1,12 @@
     using System.Reflection;
 
-    public static partial class ReflectionExtensions
+    internal static partial class ReflectionExtensions
```
Have we checked the VS repo to ensure this does not have any references that could break?

Haplois replied:
The class itself was introduced as part of FQN, never used anywhere since it didn't ship.

AbhitejJohn commented on src/Microsoft.TestPlatform.AdapterUtilities/Extensions/ReflectionExtensions.Type.cs:
```diff
@@ -1,12 +1,12 @@
     using System.Reflection;
 
-    public static partial class ReflectionExtensions
+    internal static partial class ReflectionExtensions
```
For members here as well...

Haplois replied:
done.

AbhitejJohn commented on src/Microsoft.TestPlatform.AdapterUtilities/Microsoft.TestPlatform.AdapterUtilities.csproj:
```diff
@@ -0,0 +1,53 @@
+    <TargetPlatformVersion>10.0.14393.0</TargetPlatformVersion>
+    <TargetPlatformMinVersion>10.0.10240.0</TargetPlatformMinVersion>
+    <TargetFrameworkIdentifier>.NETPortable</TargetFrameworkIdentifier>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
```
Is this required? (<TargetFrameworkVersion>v4.5</TargetFrameworkVersion>)

Haplois replied:
I actually don't know.  Without those properties, the UAP build fails. I am not an expert on that platform so just copied over from ObjectModel.

AbhitejJohn replied:
Weird, cause we are setting TFM to a different version below. So does this project target 4.5 or 4.5.1?

AbhitejJohn commented on src/Microsoft.TestPlatform.CrossPlatEngine/Microsoft.TestPlatform.CrossPlatEngine.csproj:
```diff
@@ -21,7 +21,7 @@
     <ProjectReference Include="..\Microsoft.TestPlatform.ObjectModel\Microsoft.TestPlatform.ObjectModel.csproj" />
     <ProjectReference Include="..\Microsoft.TestPlatform.Common\Microsoft.TestPlatform.Common.csproj" />
-	<ProjectReference Include="..\Microsoft.TestPlatform.Utilities\Microsoft.TestPlatform.Utilities.csproj" />
+	  <ProjectReference Include="..\Microsoft.TestPlatform.Utilities\Microsoft.TestPlatform.Utilities.csproj" />
```
nit: formatting.

Haplois replied:
fixed

AbhitejJohn commented on src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:
```diff
@@ -66,64 +63,6 @@ public TestCase(string fullyQualifiedName, Uri executorUri, string source)
-        /// <remarks>
-        /// If <paramref name="method"/> is specified, TestId will be calculated based on that instead of <paramref name="fullyQualifiedName"/>.
-        /// </remarks>
-        public TestCase(string fullyQualifiedName, MethodBase method, Uri executorUri, string source)
```
Isn't this breaking?

Haplois replied:
This was introduced as part of FQN changes. The current release doesn't have these.

AbhitejJohn commented on src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:
```diff
@@ -165,34 +104,6 @@ public Guid Id
-        ///     <code>NamespaceA.NamespaceB.ClassName`1+InnerClass`2</code>
-        /// </example>
-        [DataMember]
-        public string ManagedType { 
```
Does the serializer in TP work well with deleted members?

Haplois replied:
These members don't exist on older releases, so removed to keep compatibility.

AbhitejJohn reviewed (commented):
Concerned with the changes in OM. What does our test matrix look like?

Haplois commented on src/package/nuspec/Microsoft.TestPlatform.AdapterUtilities.nuspec:
```diff
@@ -0,0 +1,54 @@
+<package>
+    <metadata>
+        <id>Microsoft.TestPlatform.AdapterUtilities</id>
+        <version>15.0.0</version>
```
The version is overridden during the build process. It's coming from the CI. Something like `16.9.0-preview-20210106-01` for the previews and `16.8.3` for the releases.

Haplois committed:
Peer review fixes

 - Added `net20` and `net35` support to the code.
 - Other refactorings.

Haplois committed:
Peer review fixes batch 2

Haplois commented:
We're removing new FQN features from ObjectModel since they were breaking compatibility. So changes on ObjectModel should affect any older releases. They didn't ship on a release yet.  /cc: @AbhitejJohn

AbhitejJohn commented:
@Haplois : Sounds great, thanks.
